lib/spack/spack/cmd/__init__.py

```diff
@@ -279,7 +284,26 @@ def disambiguate_spec_from_hashes(spec, hashes, local=False, installed=True, fir
     if not matching_specs:
         tty.die("Spec '%s' matches no installed packages." % spec)
 
-    elif first:
+    elif best:
+        host_arch = archspec.cpu.host()
+
+        # filter out specs that won't run on this machine
+        matching_specs = [spec for spec in matching_specs if host_arch >= spec.target]
+
+        # comparators for picking best spec according to current prefs
+        ver = spack.package_prefs.PackagePrefs(spec.name, "version")
+        target = spack.package_prefs.PackagePrefs(spec.name, "target")
+        compiler = spack.package_prefs.PackagePrefs(spec.name, "compiler")
+
+        # these are harder but can be done w/package prefs:
+        # TODO: pick version with most preferred variants
+        # TODO: pick version with most preferred virtual provider (e.g. MPI)
+
+        # sort key that orders specs by prefs above
+        def sort_key(spec):
+            return (ver(spec.version), target(spec.target), compiler(spec.compiler), spec)
+
+        matching_specs.sort(key=sort_key)
```

Comment:
@adamjstewart, @haampie: see here -- it's not particularly hard to order candidates by target or version -- we already do this in the concretizer (`PackagePrefs` is what it uses).  I think this is a good enough start to merge and make default -- thoughts?